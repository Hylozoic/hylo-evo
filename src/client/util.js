export const loadScript = url => {
  var script = document.createElement('script')
  script.src = url
  const promise = new Promise((resolve, reject) => {
    script.onload = resolve
  })
  document.head.appendChild(script)
  return promise
}
