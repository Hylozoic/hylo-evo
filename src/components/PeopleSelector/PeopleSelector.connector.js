import { connect } from 'react-redux'

import getQueryParam from 'store/selectors/getQueryParam'

import {
  addParticipant,
  removeParticipant,
  fetchPeople,
  matchesSelector,
  participantsSelector,
  setAutocomplete
} from './PeopleSelector.store'
import changeQueryParam from 'store/actions/changeQueryParam'

export function getParticipantSearch (props, participantsFromStore) {
  const participants = getQueryParam('participants', null, props)
  if (participants) {
    return participants
      .split(',')
      .filter(p => !participantsFromStore.find(participant => p === participant))
  }
  return null
}

export function mapStateToProps (state, props) {
  const participants = participantsSelector(state, props)
  return {
    autocomplete: state.autocomplete,
    matches: matchesSelector(state),
    participants,
    participantSearch: getParticipantSearch(props, participants)
  }
}

export const mapDispatchToProps = {
  addParticipant,
  removeParticipant,
  changeQueryParam,
  fetchPeople,
  setAutocomplete
}

export default connect(mapStateToProps, mapDispatchToProps)
