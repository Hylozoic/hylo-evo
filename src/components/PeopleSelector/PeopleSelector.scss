.people-selector {
  display: flex;
  flex: 1;
  flex-direction: column;
}

.autocomplete-control {
  display: flex;
  flex: 1;
  border-bottom: 1px $color-rhino-30 solid;
  padding-bottom: $space-2x;
  margin-right: $space-5x;
}

.autocomplete {
  flex: 1;
  border: none;
  outline: none;

  // Yep, believe it or not this is necessary... can't group the selectors!
  &::-webkit-input-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &::-moz-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &:-ms-input-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &:-moz-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &::placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 16px;
    letter-spacing: 0.4;
    line-height: 20px;
  }
}

.close-button {
  margin-bottom: $space-4x;
  margin-left: $space-6x;
}

.thread-header {
  composes: header from 'components/Thread/Thread.scss';
  padding-bottom: $space-3x;
}

.message-form {
  composes: message-form from 'components/MessageForm/MessageForm.scss';
}
