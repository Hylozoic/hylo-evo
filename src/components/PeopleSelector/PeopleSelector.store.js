import { createSelector } from 'redux-orm'
import { pick } from 'lodash/fp'

import orm from 'store/models'

export const MODULE_NAME = 'PeopleSelector'

export const FETCH_PEOPLE = 'FETCH_PEOPLE'
export const SET_AUTOCOMPLETE = 'PeopleSelector/SET_AUTOCOMPLETE'
export const ADD_PARTICIPANT = 'PeopleSelector/ADD_PARTICIPANT'
export const REMOVE_PARTICIPANT = 'PeopleSelector/REMOVE_PARTICIPANT'

const fetchPeopleQuery =
`query PersonAutocomplete ($autocomplete: String, $first: Int) {
  people (autocomplete: $autocomplete, first: $first) {
    items {
      id
      name
      avatarUrl
      memberships {
        id
        community {
          id
          name
        }
      }
    }
  }
}`

export function fetchPeople (autocomplete, query = fetchPeopleQuery, first = 20) {
  return {
    type: FETCH_PEOPLE,
    graphql: {
      query,
      variables: { autocomplete, first }
    },
    meta: { extractModel: 'Person' }
  }
}

export function addParticipant (id) {
  return {
    type: ADD_PARTICIPANT,
    payload: id
  }
}

export function removeParticipant (id) {
  return {
    type: REMOVE_PARTICIPANT,
    payload: id
  }
}

export function setAutocomplete (autocomplete) {
  return {
    type: SET_AUTOCOMPLETE,
    payload: autocomplete
  }
}

export function participantsFromStore (state) {
  return state[MODULE_NAME].participants
}

export const participantsSelector = createSelector(
  orm,
  state => state.orm,
  participantsFromStore,
  (session, fromStore) => fromStore.map(id =>
    pick([ 'id', 'name', 'avatarUrl' ], session.Person.withId(id).ref))
)

export const matchesSelector = createSelector(
  orm,
  state => state.orm,
  state => state[MODULE_NAME].autocomplete,
  state => state[MODULE_NAME].participants,
  (session, autocomplete, participants) => {
    if (autocomplete) {
      const term = autocomplete.toLowerCase()
      const matches = session.Person
        .all()
        .filter(p => !participants.includes(p.id))
        .filter(p => p.name.toLowerCase().includes(term))
        .orderBy('name')
      return matches.toModelArray().map(match => ({
        ...pick([ 'id', 'name', 'avatarUrl' ], match.ref),
        community: match.memberships.first()
          ? match.memberships.first().community.name : null
      }))
    }
    return null
  }
)

export const defaultState = {
  participants: []
}

export default function reducer (state = defaultState, action) {
  const { error, payload, type } = action
  if (error) return state

  switch (type) {
    case SET_AUTOCOMPLETE:
      return {
        ...state,
        autocomplete: payload
      }

    case ADD_PARTICIPANT:
      return {
        ...state,
        participants: [ ...state.participants, payload ]
      }

    case REMOVE_PARTICIPANT:
      if (payload) {
        return {
          ...state,
          participants: state.participants.filter(p => p !== payload)
        }
      }
      // No payload? Remove the last element.
      return {
        ...state,
        participants: state.participants.slice(0, state.participants.length - 1)
      }
  }

  return state
}
