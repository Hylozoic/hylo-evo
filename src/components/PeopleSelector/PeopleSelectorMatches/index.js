import PeopleSelectorMatches from './PeopleSelectorMatches'
export default PeopleSelectorMatches
