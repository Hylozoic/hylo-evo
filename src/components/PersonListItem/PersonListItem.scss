.person-list-item {
  list-style: none;
  display: flex;
  align-items: center;
  justify-content: center;
  max-width: 500px;
  margin: $space-1x 0;
  padding: $space-2x;
  border-radius: $space-1x;
  cursor: pointer;
}

.active {
  background-color: $color-rhino-10;
}

.avatar {
  margin-right: $space-4x;
  border: none;
}

.name {
  composes: bdy-lt-sm from 'css/typography.scss';
  flex: 1;
  padding-right: $space-2x;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.community {
  composes: faint-lt-sm from 'css/typography.scss';
}
