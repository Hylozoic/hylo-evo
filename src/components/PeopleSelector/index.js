import PeopleSelector from './PeopleSelector'
import connector from './PeopleSelector.connector'
export default connector(PeopleSelector)
