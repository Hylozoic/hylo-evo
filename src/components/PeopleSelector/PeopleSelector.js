import PropTypes from 'prop-types'
import React from 'react'
import { debounce, throttle } from 'lodash/fp'

import { getKeyCode, keyMap } from 'util/textInput'
import MessageForm from 'components/MessageForm'
import CloseMessages from 'components/Thread/CloseMessages'
import PeopleSelectorMatches from './PeopleSelectorMatches'
import PeopleSelectorContacts from './PeopleSelectorContacts'
import SelectorMatchedItem from './SelectorMatchedItem'
import './PeopleSelector.scss'

const { any, arrayOf, func, shape, string } = PropTypes

const personType = shape({
  id: any,
  name: string,
  avatarUrl: string
})

// TODO: This _grossly_ underestimates the problem! See:
// https://www.w3.org/International/questions/qa-personal-names
const invalidPersonName = /[^a-z '-]+/gi

export default class PeopleSelector extends React.Component {
  static propTypes = {
    autocomplete: string,
    fetchPeople: func,
    deleteParticipant: func,
    participants: arrayOf(personType),
    setAutocomplete: func
  }

  constructor (props) {
    super(props)
    this.state = { currentMatch: null }
  }

  componentDidMount () {
    this.props.fetchContacts()
    this.props.fetchRecentContacts()
    const { participantSearch } = this.props
    if (participantSearch) {
      participantSearch.forEach(p => this.props.addParticipant(p))
      this.props.changeQueryParam(this.props, 'participants', null)
    }
  }

  componentWillReceiveProps (props) {
    const { matches } = props
    if (!matches || matches.length === 0) {
      this.setState({ currentMatch: null })
      return
    }

    if (matches.find(m => m.id === this.state.currentMatch)) return
    this.setState({ currentMatch: matches[0].id })
  }

  addParticipant = id => {
    this.props.setAutocomplete(null)
    this.setState({ currentMatch: null })
    this.autocomplete.value = null
    this.props.addParticipant(id)
    this.autocomplete.focus()
  }

  arrow (direction, evt) {
    evt.preventDefault()
    let delta = 0
    const idx = this.props.matches.findIndex(m => m.id === this.state.currentMatch)
    if (direction === 'up') {
      if (idx > 0) delta = -1
    }
    if (direction === 'down') {
      if (idx < this.props.matches.length - 1) delta = 1
    }
    this.setCurrentMatch(this.props.matches[idx + delta].id)
  }

  autocompleteSearch = throttle(1000, this.props.fetchPeople)

  onChange = debounce(200, () => {
    const { value } = this.autocomplete
    if (!invalidPersonName.exec(value)) {
      return this.props.setAutocomplete(value)
    }
    this.autocomplete.value = value.replace(invalidPersonName, '')
  })

  onKeyDown (evt) {
    switch (getKeyCode(evt)) {
      case keyMap.BACKSPACE: return this.state.currentMatch ? null : this.props.removeParticipant()
      case keyMap.UP: return this.arrow('up', evt)
      case keyMap.DOWN: return this.arrow('down', evt)
      case keyMap.COMMA:
      case keyMap.ENTER:
        evt.preventDefault()
        this.props.addParticipant(this.state.currentMatch)
        this.autocomplete.value = null
        return this.props.setAutocomplete(null)

      default:
        this.autocompleteSearch(this.autocomplete.value)
    }
  }

  removeParticipant (id) {
    this.props.removeParticipant(id)
    if (!this.autocomplete.value) {
      this.setState({ currentMatch: null })
    }
  }

  setCurrentMatch = id => this.setState({ currentMatch: id })

  render () {
    const { contacts, matches, participants, recentContacts, removeParticipant, findOrCreateThread } = this.props
    const { currentMatch } = this.state
    return <div styleName='people-selector'>
      <div styleName='thread-header' tabIndex='0'>
        <div styleName='autocomplete-control'>
          {participants && participants.map(participant =>
            <SelectorMatchedItem
              avatarUrl={participant.avatarUrl}
              key={participant.id}
              name={participant.name}
              removeParticipant={() => removeParticipant(participant.id)} />
          )}
          <input styleName='autocomplete'
            autoFocus
            ref={i => this.autocomplete = i} // eslint-disable-line no-return-assign
            type='text'
            spellCheck={false}
            onChange={evt => this.onChange(evt)}
            onKeyDown={evt => this.onKeyDown(evt)}
            placeholder={participants.length ? '' : 'Type in the names of people to message'} />
        </div>
        <CloseMessages />
      </div>
      {currentMatch
        ? <PeopleSelectorMatches
          addParticipant={this.addParticipant}
          currentMatch={currentMatch}
          matches={matches}
          setCurrentMatch={this.setCurrentMatch} />
        : <PeopleSelectorContacts
          addParticipant={this.addParticipant}
          contacts={contacts}
          recentContacts={recentContacts} />}
      {participants && participants.length > 0 &&
        <div styleName='message-form'>
          <MessageForm ref='form'
            forNewThread
            findOrCreateThread={() => findOrCreateThread(participants.map(p => p.id))} />
        </div>}
    </div>
  }
}
