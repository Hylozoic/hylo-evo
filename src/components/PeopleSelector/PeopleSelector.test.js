import { mount, shallow, render } from 'enzyme'
import { MemoryRouter } from 'react-router'
import React from 'react'

import { keyMap } from 'util/textInput'
import PeopleSelector from './PeopleSelector'
import PersonListItem from 'components/PersonListItem'

describe('PeopleSelector', () => {
  it('matches the last snapshot', () => {
    const wrapper = shallow(
      <PeopleSelector fetchPeople={() => {}} matches={[]} participants={[]} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  describe('onKeyDown', () => {
    let addParticipant
    let fetchPeople
    let removeParticipant
    let input
    let setAutocomplete
    let wrapper

    beforeEach(() => {
      addParticipant = jest.fn()
      fetchPeople = jest.fn()
      removeParticipant = jest.fn()
      setAutocomplete = jest.fn()
      wrapper = mount(
        <MemoryRouter>
          <PeopleSelector
            addParticipant={addParticipant}
            matches={[ { id: '1' }, { id: '2' } ]}
            participants={[]}
            fetchPeople={fetchPeople}
            removeParticipant={removeParticipant}
            setAutocomplete={setAutocomplete} />
        </MemoryRouter>
      )
      wrapper.find(PeopleSelector).node.setState({ currentMatch: '1' })
      input = wrapper.find('input').first()
    })

    it('does not hit server when backspace is pressed', () => {
      input.simulate('keyDown', { keyCode: keyMap.BACKSPACE })
      expect(fetchPeople).not.toHaveBeenCalled()
    })

    it('hits server when keys other than backspace are pressed', () => {
      input.simulate('keyDown', { keyCode: keyMap.SPACE })
      expect(fetchPeople).toHaveBeenCalled()
    })

    it('removes participant if backspace pressed when currentMatch missing', () => {
      wrapper.find(PeopleSelector).node.setState({ currentMatch: null })
      input.simulate('keyDown', { keyCode: keyMap.BACKSPACE })
      expect(removeParticipant).toHaveBeenCalled()
    })

    it('does not remove paraticipant if backspace pressed when currentMatch defined', () => {
      input.simulate('keyDown', { keyCode: keyMap.BACKSPACE })
      expect(removeParticipant).not.toHaveBeenCalled()
    })

    it('calls arrow with `up` if up arrow pressed', () => {
      const arrow = PeopleSelector.prototype.arrow
      PeopleSelector.prototype.arrow = jest.fn()
      input.simulate('keyDown', { keyCode: keyMap.UP })
      const calls = PeopleSelector.prototype.arrow.mock.calls
      expect(calls[calls.length -1][0]).toBe('up')
      PeopleSelector.prototype.arrow = arrow
    })

    it('calls arrow with `down` if down arrow pressed', () => {
      const arrow = PeopleSelector.prototype.arrow
      PeopleSelector.prototype.arrow = jest.fn()
      input.simulate('keyDown', { keyCode: keyMap.DOWN })
      const calls = PeopleSelector.prototype.arrow.mock.calls
      expect(calls[calls.length -1][0]).toBe('down')
      PeopleSelector.prototype.arrow = arrow
    })

    it('does not change active match if at top of list when up arrow pressed', () => {
      input.simulate('keyDown', { keyCode: keyMap.UP })
      const actual = wrapper.find(PersonListItem).first().prop('active')
      expect(actual).toBe(true)
    })

    it('changes active match if not at top of list when up arrow pressed', () => {
      wrapper.find(PeopleSelector).node.setState({ currentMatch: '2' })
      input.simulate('keyDown', { keyCode: keyMap.UP })
      const actual = wrapper.find(PersonListItem).last().prop('active')
      expect(actual).toBe(false)
    })

    it('does not change active match if at bottom of list when down arrow pressed', () => {
      wrapper.find(PeopleSelector).node.setState({ currentMatch: '2' })
      input.simulate('keyDown', { keyCode: keyMap.DOWN })
      const actual = wrapper.find(PersonListItem).last().prop('active')
      expect(actual).toBe(true)
    })

    it('changes active match if not at bottom of list when down arrow pressed', () => {
      input.simulate('keyDown', { keyCode: keyMap.DOWN })
      const actual = wrapper.find(PersonListItem).first().prop('active')
      expect(actual).toBe(false)
    })

    it('calls addParticipant with currentMatch when enter pressed', () => {
      input.simulate('keyDown', { keyCode: keyMap.ENTER })
      expect(addParticipant).toBeCalledWith('1')
    })

    it('calls addParticipant with currentMatch when comma pressed', () => {
      input.simulate('keyDown', { keyCode: keyMap.COMMA })
      expect(addParticipant).toBeCalledWith('1')
    })

    it('resets values after adding participants', () => {
      input.simulate('keyDown', { keyCode: keyMap.ENTER })
      expect(setAutocomplete).toBeCalledWith(null)
      expect(input.node.value).toBe('')
    })
  })

  describe('addParticipant', () => {
    it('calls addParticipant with the correct id', () => {
      const addParticipant = jest.fn()
      const wrapper = mount(
        <MemoryRouter>
          <PeopleSelector
            addParticipant={addParticipant}
            participants={[]}
            fetchPeople={() => {}}
            setAutocomplete={() => {}} />
        </MemoryRouter>
      )
      wrapper.find(PeopleSelector).node.addParticipant('1')
      expect(addParticipant).toBeCalledWith('1')
    })

    it('resets values after adding a participant', () => {
      const setAutocomplete = jest.fn()
      const wrapper = mount(
        <MemoryRouter>
          <PeopleSelector
            addParticipant={() => {}}
            participants={[]}
            fetchPeople={() => {}}
            setAutocomplete={setAutocomplete} />
        </MemoryRouter>
      )
      const input = wrapper.find('input').first()
      input.node.value = 'flargle'
      wrapper.find(PeopleSelector).node.addParticipant('1')
      expect(input.node.value).toBe('')
      expect(setAutocomplete).toBeCalledWith(null)
    })
  })

  describe('componentDidMount', () => {
    it('adds particpants in the search, then clears it', () => {
      const addParticipant = jest.fn()
      const changeQueryParam = jest.fn()
      const wrapper = mount(
        <MemoryRouter>
          <PeopleSelector
            addParticipant={addParticipant}
            participants={[]}
            participantSearch={[ '1', '2' ]}
            fetchPeople={() => {}}
            changeQueryParam={changeQueryParam} />
        </MemoryRouter>
      )
      expect(addParticipant).toBeCalledWith('1')
      expect(addParticipant).toBeCalledWith('2')
      const [ _, param, value ] = changeQueryParam.mock.calls[0]
      expect(param).toBe('participants')
      expect(value).toBe(null)
    })
  })

  describe('setAutocomplete', () => {
    let setAutocomplete
    let wrapper

    beforeEach(() => {
      jest.useFakeTimers()
      setAutocomplete = jest.fn()
      wrapper = mount(
        <MemoryRouter>
          <PeopleSelector
            fetchPeople={() => {}}
            participants={[]}
            setAutocomplete={setAutocomplete} />
        </MemoryRouter>
      )
    })

    it('updates if user input contains valid characters', () => {
      const expected = 'Poor Yorick'
      const input = wrapper.find('input').first()
      input.node.value = expected
      input.simulate('change')
      jest.runAllTimers()
      const actual = setAutocomplete.mock.calls[0][0]
      expect(actual).toBe(expected)
    })

    it('does not update if user input contains invalid characters', () => {
      const invalid = 'Poor Yorick9238183$@#$$@!'
      const expected = 'Poor Yorick'
      const input = wrapper.find('input').first()
      input.node.value = invalid
      input.simulate('change')
      jest.runAllTimers()
      expect(setAutocomplete).not.toHaveBeenCalled()
      expect(input.node.value).toBe(expected)
    })
  })
})
