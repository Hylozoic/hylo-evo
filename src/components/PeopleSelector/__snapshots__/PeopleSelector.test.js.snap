// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleSelector matches the last snapshot 1`] = `
<div
  data-styleName="people-selector"
>
  <div
    data-styleName="thread-header"
    tabIndex="0"
  >
    <div
      data-styleName="autocomplete-control"
    >
      <input
        autoFocus={true}
        data-styleName="autocomplete"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Type in the names of people to message"
        spellCheck={false}
        type="text"
      />
    </div>
    <CloseMessages />
  </div>
  <PeopleSelectorContacts />
</div>
`;
